### PR TITLE
[SPARK-15718][SQL] better error message for writing bucketed data

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently we don't support bucketing for `save` and `insertInto`.

For `save`, we just write the data out into a directory users specified, and it's not a table, we don't keep its metadata. When we read it back, we have no idea if the data is bucketed or not, so it doesn't make sense to use `save` to write bucketed data, as we can't use the bucket information anyway.

We can support it in the future, once we have features like bucket discovery, or we save bucket information in the data directory too, so that we don't need to rely on a metastore.

For `insertInto`, it inserts data into an existing table, so it doesn't make sense to specify bucket information, as we should get the bucket information from the existing table.

This PR improves the error message for the above 2  cases.
## How was this patch tested?

new test in `BukctedWriteSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @andrewor14 @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -500,10 +500,10 @@ final class DataFrameWriter private[sql](df: DataFrame) {
         s"existing columns (${validColumnNames.mkString(", ")})"))
   }
 
-  private def assertNotBucketed(): Unit = {
+  private def assertNotBucketed(operation: String): Unit = {
     if (numBuckets.isDefined || sortColumnNames.isDefined) {
       throw new IllegalArgumentException(
-        "Currently we don't support writing bucketed data to this data source.")
+        s"'$operation' does not support bucketing right now.")
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: It looks weird to pass in the operation as a string. Instead you can get the operation from the caller of `assertNotBucketed`, something like this: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala#L103

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -502,8 +503,18 @@ final class DataFrameWriter private[sql](df: DataFrame) {
 
   private def assertNotBucketed(): Unit = {
     if (numBuckets.isDefined || sortColumnNames.isDefined) {
+      val ourMethodName = "assertNotBucketed"
+      val callerMethodName = Thread.currentThread.getStackTrace()
```
Comment (tejasapatil) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I see similar stuff done here to get the caller name : https://github.com/apache/spark/blob/master/streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala#L262

This logic could be moved to a util method. Utils class seems to have some method regarding getting callsite : https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/Utils.scala#L1331

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedWriteSuite.scala [hunk 3]
```diff
@@ -59,11 +59,18 @@ class BucketedWriteSuite extends QueryTest with SQLTestUtils with TestHiveSingle
     intercept[SparkException](df.write.bucketBy(3, "i").format("text").saveAsTable("tt"))
   }
 
-  test("write bucketed data to non-hive-table or existing hive table") {
+  test("write bucketed data using save() or insertInto()") {
```
Comment (tejasapatil) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedWriteSuite.scala]: Ideally each test case could be atomic. You could split this test case into two parts.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: we can fix that later, maybe add a TODO or something

**Comment by andrewor14** [comment]
@cloud-fan actually since this is only used in 2 places maybe you can just inline it. It's probably not worth the trouble to traverse the call stack to find the method name.

**Comment by andrewor14** [comment]
This looks OK. Merging into master 2.0.

**Comment by Downchuck** [comment]
Regarding the reason for disallowing bucket writes: "we have no idea [on read] if the data is bucketed or not, so it doesn't make sense to use save to write bucketed data"

It's easy enough to pass information to the reader, it doesn't need to be automatic or rely on a metastore or other discover methods. Something as simple as read.sortedBy(cols...).bucketedBy(func Or cols..) would do.

**Pull request closed without merging**